hkmc2/shared/src/main/scala/hkmc2/codegen/Lifter.scala

```diff
@@ -0,0 +1,1504 @@
+        case _ => N
+      case _ => N
+  
+  object InstSel:
+    def unapply(p: Path) = p match
+      case Value.Ref(l: BlockMemberSymbol) => S(l)
+      case s @ Select(Value.Ref(l: BlockMemberSymbol), Tree.Ident("class")) => S(l)
+      case _ => N
+
+  def modOrObj(d: Defn) = d match
+    case c: ClsLikeDefn => (c.k is syntax.Mod) || (c.k is syntax.Obj)
+    case _ => false
+
+  // TODO: this is very unclean
+  def isContClassPth(p: Path)(using s: State) = p match
+    case Select(Select(Select(Value.Ref(s.globalThisSymbol), Tree.Ident("Predef")), 
+      Tree.Ident("__Cont")), Tree.Ident("class")) => true
+    case _ => false
+
+
+/**
+  * Lifts classes and functions to the top-level. Also automatically rewrites lambdas.
+  * Assumes the input block does not have any `HandleBlock`s.
+  */
+class Lifter(using State, Raise):
+  import Lifter.*
+
+  /**
+    * The context of the class lifter. One can create an empty context using `Lifter.empty`.
+    * 
+    * @param defns A map from all BlockMemberSymbols to their definitions.
+    * @param defnsCur All definitions that are nested in the current top level definition.
+    * @param nestedDefns Definitions which are nested in a given definition (shallow).
+    * @param usedLocals Describes the locals belonging to each function that are accessed/mutated by nested definitions.
+    * @param accessInfo Which previously defined variables/definitions could be accessed/modified by a particular definition, 
+    * possibly through calls to other functions or by constructing a class.
+    * @param ignoredDefns The definitions which must not be lifted.
+    * @param inScopeDefns Definitions which are in scope to another definition (excluding itself and its nested definitions).
+    * @param modLocals A map from the modules and objects to the local to which it is instantiated after lifting.
+    * @param localCaptureSyms The symbols in a capture corresponding to a particular local
+    * @param prevFnLocals Locals belonging to function definitions that have already been traversed
+    * @param prevClsDefns Class definitions that have already been traversed, excluding modules
+    * @param curModules Modules that that we are currently nested in (cleared if we are lifted out)
+    * @param capturePaths The path to access a particular function's capture in the local scope
+    * @param bmsReqdInfo The (mutable) captures and (immutable) local variables each function requires
+    * @param ignoredBmsPaths The path to access a particular BlockMemberSymbol (for definitions which could not be lifted)
+    * @param localPaths The path to access a particular local (possibly belonging to a previous function) in the current scope
+    * @param iSymPaths The path to access a particular `innerSymbol` (possibly belonging to a previous class) in the current scope
+    * @param replDefns Ignored definitions whose definitions we should replace in the block itself
+    */
+  case class LifterCtx(
```

Comment:
This class is not meant to be instantiated directly except in `empty`. Doing so would be dangerous because of all the default arguments. So make it
```suggestion
  case class LifterCtx private(
```